### PR TITLE
feat(api): add GraphQL downloadEvidence endpoint

### PR DESCRIPTION
**Pull request by marcpfuller**
### Issue

This PR addresses issue #722 #599 

### Description of the Change

This PR adds a new GraphQL action endpoint `downloadEvidence` that allows authenticated users to download evidence files via the API. The endpoint allows users to retrieve either:

1. `downloadUrl` : Returns a URL to the evidence download endpoint
2. `fileBase64`: Returns the file content as base64-encoded data along with metadata

The implementation leverages the existing `JwtRequiredMixin` and `HasuraActionView` base classes to handle:
- JWT token authentication
- Hasura action secret validation
- Input validation (requires `id` field)
- Consistent error response formatting

**Permission Model:**
- Users with "manager" role can access all evidence files
- Regular users can only access evidence from projects they're assigned to

**Key Components:**
- New `GraphqlDownloadEvidence` view in `ghostwriter/api/views.py`
- URL route added to `ghostwriter/api/urls.py`
- Hasura action definition in `hasura-docker/metadata/actions.graphql`
- Hasura permissions configured in `hasura-docker/metadata/actions.yaml`
- Comprehensive test suite with 12 test cases covering authentication, authorization, success scenarios, and edge cases

### Alternate Designs

**Alternative 1: Direct file streaming**
- Considered streaming the file directly from the API endpoint instead of returning URLs
- Rejected because it would require handling large file transfers in the API layer and doesn't leverage existing evidence download infrastructure

**Alternative 2: Separate endpoints for URL vs base64**
- Considered creating two separate actions: `downloadEvidence` and `downloadEvidenceBase64`
- Rejected because a single endpoint with an optional parameter is more flexible and reduces code duplication

**Alternative 3: Using API keys instead of JWT**
- Considered supporting API key authentication
- Rejected to maintain consistency with other GraphQL actions that use JWT tokens

### Possible Drawbacks

1. **Base64 encoding overhead**: When `returnBase64: true`, large files will consume significant memory and bandwidth. However, this is an optional feature and the URL mode is the default.

2. **File system dependency**: The endpoint checks `os.path.exists()` which assumes evidence files are on the local filesystem. This could be an issue in distributed deployments, but matches the existing pattern used throughout Ghostwriter.

3. **No file size limits**: The endpoint doesn't enforce size limits on base64 encoding. Very large files could cause memory issues, though this is mitigated by the fact that evidence files are typically screenshots and small documents.

### Verification Process

**Test Coverage:**
- Added 12 comprehensive test cases covering:
  - Authentication failures (missing token, invalid token, missing action secret, wrong action secret)
  - Input validation (missing evidence ID)
  - Authorization checks (project access control, manager bypass)
  - Success scenarios (evidence with finding, evidence with report, manager access)
  - File operations (base64 encoding, file not found handling)
  - Request handling (forwarded IP headers)

**Manual Testing:**
- Tested via curl commands with valid JWT tokens
- Tested via  GraphQL query 
- Verified download URL generation returns correct paths
- Verified base64 encoding returns decodable file content
- Verified permission checks block unauthorized access
- Verified manager users can access all projects
- Verified regular users can only access assigned projects

**Curl Command**
```
# Login to get a JWT token
curl -X POST http://localhost:8080/v1/graphql \
  -H "Content-Type: application/json" \
  -H "x-hasura-admin-secret: <admin secret>" \
  -d '{
    "query": "mutation Login($username: String!, $password: String!) { login(username: $username, password: $password) { token user { id username role } } }",
    "variables": {
      "username": "your_username",
      "password": "your_password"
    }
  }'

# Gets download link
# Request both downloadUrl and fileBase64
curl -X POST http://localhost:8080/v1/graphql \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer <JWT TOKEN>" \
  -H "Hasura-Action-Secret: <ACTION SECRET>" \
  -d '{
    "query": "mutation DownloadEvidence($evidenceId: Int!) { downloadEvidence(evidenceId: $evidenceId) { evidenceId filename friendlyName downloadUrl } }",
    "variables": {
      "evidenceId": 1
    }
  }'

# Download evidence as base64
# Request both downloadUrl and fileBase64
curl -X POST http://localhost:8080/v1/graphql \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer <JWT TOKEN>" \
  -H "Hasura-Action-Secret: <ACTION SECRET>" \
  -d '{
    "query": "mutation DownloadEvidence($evidenceId: Int!) { downloadEvidence(evidenceId: $evidenceId) { evidenceId filename friendlyName fileBase64 } }",
    "variables": {
      "evidenceId": 1
    }
  }'

```
***GraphQL Query***
```
# Get both download url or filebase64
mutation GetDownloadUrl($evidenceId: Int!) {
  downloadEvidence(evidenceId: $evidenceId) {
    evidenceId
    filename
    friendlyName
    downloadUrl
    fileBase64
  }
}
```

**Integration Testing:**
- All existing tests pass
- New endpoint integrates seamlessly with existing authentication infrastructure
- Error responses follow Hasura action error format conventions

### Release Notes

Added GraphQL `downloadEvidence` action endpoint for programmatic access to evidence files with JWT authentication and role-based permissions

### CODE REVIEW

File: ghostwriter/api/views.py [hunk 1]
```diff
@@ -577,6 +582,91 @@ def post(self, request, *args, **kwargs):
...
+        project = evidence.finding.report.project if evidence.finding else evidence.report.project
+
+        # Managers can access all evidence, others need project assignment
+        if self.user_obj.role != "manager":
+            # Check if user has access to the project
+            if not ProjectAssignment.objects.filter(project=project, operator=self.user_obj).exists():
+                return JsonResponse(
+                    utils.generate_hasura_error_payload("Forbidden - no access to project", "Unauthorized"),
+                    status=HTTPStatus.FORBIDDEN
+                )
```
Comment (ColonelThirtyTwo) [inline ghostwriter/api/views.py]: I've been adding `user_can_view` and `user_viewable` methods to models to avoid duplicating the authorization logic every time we access a model. Can you add these methods to Evidence?

File: ghostwriter/api/views.py [hunk 2]
```diff
@@ -577,6 +582,91 @@ def post(self, request, *args, **kwargs):
...
+
+        # Check if file exists
+        if not os.path.exists(evidence.document.path):
```
Comment (ColonelThirtyTwo) [inline ghostwriter/api/views.py]: Catch the not found error from reading the file rather than checking beforehand.

File: ghostwriter/api/views.py [hunk 3]
```diff
@@ -577,6 +582,91 @@ def post(self, request, *args, **kwargs):
...
+            logger.error("Evidence file not found: %s", evidence.document.path)
+            return JsonResponse(
+                utils.generate_hasura_error_payload("Evidence file not found on server", "EvidenceFileNotFound"),
+                status=HTTPStatus.NOT_FOUND
+            )
+
+        # Return response based on request type
+        if return_base64:
+            try:
+                with open(evidence.document.path, 'rb') as f:
```
Comment (ColonelThirtyTwo) [inline ghostwriter/api/views.py]: `evidence.document` is a file like object - you should be able to read it directly rather than needing to open it again.

File: ghostwriter/api/views.py [hunk 4]
```diff
@@ -577,6 +582,91 @@ def post(self, request, *args, **kwargs):
...
+                    file_data = f.read()
+                    encoded_data = b64encode(file_data).decode('utf-8')
+
+                return JsonResponse({
+                    "id": evidence.id,
+                    "filename": evidence.filename,
+                    "friendlyName": evidence.friendly_name,
+                    "fileBase64": encoded_data,
+                    "contentType": mimetypes.guess_type(evidence.document.path)[0] or "application/octet-stream"
```
Comment (ColonelThirtyTwo) [inline ghostwriter/api/views.py]: I don't know if I'd even bother guessing the mime since it doesn't add any info that the consumer couldn't guess by itself. If we do keep this, we may want to switch to `guess_file_type` if we're on Python 3.13 or newer.

File: ghostwriter/api/views.py [hunk 5]
```diff
@@ -577,6 +582,91 @@ def post(self, request, *args, **kwargs):
...
+                })
+            except Exception as e:
+                logger.exception("Failed to encode evidence file: %s", e)
+                return JsonResponse(
+                    utils.generate_hasura_error_payload("Failed to read evidence file", "FileReadError"),
+                    status=HTTPStatus.INTERNAL_SERVER_ERROR
+                )
+        else:
+            # Return download URL
+            download_url = request.build_absolute_uri(
```
Comment (ColonelThirtyTwo) [inline ghostwriter/api/views.py]: I can't find the documentation for `urllib3.request.build_absolute_uri` anywhere - what does this do? Do we need to provide the hostname as well?

File: ghostwriter/api/tests/test_views.py [hunk 6]
```diff
@@ -2665,3 +2669,378 @@ def test_get_admin_results(self):
...
+            self.assertIn("contentType", result)
+
+            # Verify base64 content
+            decoded_content = base64.b64decode(result["fileBase64"])
+            self.assertEqual(decoded_content, test_content)
+
+        finally:
+            # Clean up: delete the file
+            if self.evidence_with_finding.document:
+                self.evidence_with_finding.document.delete(save=False)
```
Comment (ColonelThirtyTwo) [inline ghostwriter/api/tests/test_views.py]: Do we need to do this? I feel like django's test harness should be cleaning up any files that the test makes.

**Review by ColonelThirtyTwo — CHANGES_REQUESTED** [review]

File: ghostwriter/api/views.py (hunk 5 above)
Comment (marcpfuller) [reply ghostwriter/api/views.py]: @ColonelThirtyTwo From my understanding urllib3.request is not being used in this module. It django looks to be using HTTPRequest
https://docs.djangoproject.com/en/5.2/ref/request-response/

Comment (marcpfuller) [reply ghostwriter/api/views.py]: The hostname does not need to be provided

Comment (ColonelThirtyTwo) [reply ghostwriter/api/views.py]: I see. The `from urllib3 import request` import was confusing me.

File: ghostwriter/api/views.py [hunk 7]
```diff
@@ -577,6 +581,89 @@ def post(self, request, *args, **kwargs):
...
+        # Check if user has permission to view the evidence
+        project = evidence.finding.report.project if evidence.finding else evidence.report.project
+        if not project.user_can_view(self.user_obj):
+            return JsonResponse(
+            utils.generate_hasura_error_payload("Forbidden - no access to project", "Unauthorized"),
+            status=HTTPStatus.FORBIDDEN
+            )
+
+        # Return response based on request type
+        if return_base64:
```
Comment (chrismaddalena) [inline ghostwriter/api/views.py]: It's possible this may not be included or may be set `false` when the user has requested `filebase64` in their output. We should ensure the base64 is returned if it is requested (without needing an extra input).

File: ghostwriter/api/views.py [hunk 8]
```diff
@@ -577,6 +581,89 @@ def post(self, request, *args, **kwargs):
...
+            return JsonResponse(
+                utils.generate_hasura_error_payload("Evidence not found", "EvidenceDoesNotExist"),
+                status=HTTPStatus.NOT_FOUND
+            )
+
+        # Check if user has permission to view the evidence
+        project = evidence.finding.report.project if evidence.finding else evidence.report.project
+        if not project.user_can_view(self.user_obj):
+            return JsonResponse(
+            utils.generate_hasura_error_payload("Forbidden - no access to project", "Unauthorized"),
```
Comment (chrismaddalena) [inline ghostwriter/api/views.py]: Message should be consistent with the other endpoints and return `Unauthorized access` like this: `utils.generate_hasura_error_payload("Unauthorized access", "Unauthorized")`

File: ghostwriter/api/views.py [hunk 9]
```diff
@@ -577,6 +581,89 @@ def post(self, request, *args, **kwargs):
...
+                )
+            except Exception as e:
+                logger.exception("Unexpected error encoding evidence file: %s", e)
+                return JsonResponse(
+                    utils.generate_hasura_error_payload("Failed to encode evidence file", "FileEncodeError"),
+                    status=HTTPStatus.INTERNAL_SERVER_ERROR
+                )
+        else:
+            # Return download URL
+            download_url = request.build_absolute_uri(
```
Comment (chrismaddalena) [inline ghostwriter/api/views.py]: This returns `django:8000` in the URL, but that will only work from within a container. We should return a download link using the configured hostname from the general settings + port. We could also return just the URI for someone to append to their Ghsotwriter instance's protocol, host, and port.

File: hasura-docker/metadata/actions.graphql [hunk 10]
```diff
@@ -254,3 +254,21 @@ type GetOplogEntryByTagsResponse {
...
+type DownloadEvidenceOutput {
+  id: Int!
+  filename: String!
+  friendlyName: String!
+  downloadUrl: String
+  fileBase64: String
+  contentType: String
+}
+
+input DownloadEvidenceInput {
```
Comment (chrismaddalena) [inline hasura-docker/metadata/actions.graphql]: I think we can drop this altogether and always return the `downloadUrl` and `fileBase64` values. Hasura will handle providing the values to the user based on their requested return values. This also simplifies the issue with the required inputs test not passing. The required input is set to `id`, but the actual input from this is `input` with an `id` value in the dict. We can change the action to:

```
type Query {
  downloadEvidence(evidenceId: Int!): DownloadEvidenceOutput
}
```
